### PR TITLE
Task branch: files in tasks added with test

### PR DESCRIPTION
**Pull request by Sofia-Ingl**

### CODE REVIEW

**Comment by IP-13** [comment]
|Overall Project|70.3% **`-0.83%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|76.47%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[TaskController.java](https://github.com/itmo-events-app/event-app/blob/821540361933fabd5d16214b80906d89e5b585dc/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FTaskController.java)|91.32%|:green_apple:|
|[JwtTokenUtil.java](https://github.com/itmo-events-app/event-app/blob/821540361933fabd5d16214b80906d89e5b585dc/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Futil%2FJwtTokenUtil.java)|90.7% **`-1.16%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/821540361933fabd5d16214b80906d89e5b585dc/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|87.43% **`-6.97%`**|:green_apple:|

**Comment by IP-13** [comment]
|Overall Project|70.3% **`-0.83%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|76.47%|:green_apple:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[TaskController.java](https://github.com/itmo-events-app/event-app/blob/722dd618d6bd03cd32c997363bdef7273c0a3770/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FTaskController.java)|91.32%|:green_apple:|
|[JwtTokenUtil.java](https://github.com/itmo-events-app/event-app/blob/722dd618d6bd03cd32c997363bdef7273c0a3770/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Futil%2FJwtTokenUtil.java)|90.7% **`-1.16%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/722dd618d6bd03cd32c997363bdef7273c0a3770/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|87.43% **`-6.97%`**|:green_apple:|

File: src/test/resources/db/test-migration/V0_2__create_tables_test.sql [hunk 1]
```diff
@@ -93,21 +93,21 @@ create table if not exists event
     participant_age_highest int,
     preparing_start timestamp,
     preparing_end timestamp
-);
+    );
 create table if not exists participant(
-    id int generated always as identity primary key not null,
-    name varchar(256) not null ,
+                                          id int generated always as identity primary key not null,
```
[the remark this answers is not in the document]
Comment (IP-13) [reply src/test/resources/db/test-migration/V0_2__create_tables_test.sql:98]: я про пробелы всякие

File: src/main/java/org/itmo/eventapp/main/model/entity/Task.java [hunk 2]
```diff
@@ -47,4 +48,7 @@ public class Task {
     private LocalDateTime deadline;
 
     private LocalDateTime reminder;
+
+    @OneToMany(mappedBy = "task")
+    List<TaskObject> taskObjects;
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/model/entity/Task.java:53]: Го уберем это, хибернейт слишком много запросов делает иначе, на каждое обращение к таске будет файлы вытягивать

File: src/main/java/org/itmo/eventapp/main/service/TaskService.java [hunk 3]
```diff
@@ -124,10 +133,78 @@ public Task edit(Integer id, TaskRequest taskRequest) {
         return newTaskData;
     }
 
+    @Transactional
     public void delete(Integer id) {
+        Task task = taskRepository.findById(id).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, ExceptionConst.TASK_NOT_FOUND_MESSAGE));
+        List<Integer> taskObjectIds = new ArrayList<>();
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/service/TaskService.java:139]: МБ тут лучше taskObjectRepository.findByTaskId и по ним удалять

File: src/main/java/org/itmo/eventapp/main/service/TaskService.java [hunk 4]
```diff
@@ -124,10 +133,78 @@ public Task edit(Integer id, TaskRequest taskRequest) {
...
         taskRepository.deleteById(id);
     }
 
+
+    @Transactional
+    public List<TaskObject> addFiles(Integer id, List<MultipartFile> files) {
+
+        Task task = taskRepository.findById(id).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, ExceptionConst.TASK_NOT_FOUND_MESSAGE));
+
+        List<TaskObject> taskObjects = new ArrayList<>();
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/service/TaskService.java:159]: то же, что и выше

File: src/test/resources/db/test-migration/V0_2__create_tables_test.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (Sofia-Ingl) [reply src/test/resources/db/test-migration/V0_2__create_tables_test.sql:98]: эээээээ хз как оно образовалось ахаха

**Comment by IP-13** [comment]
|Overall Project|65.91% **`-4.26%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|30.75%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[TaskController.java](https://github.com/itmo-events-app/event-app/blob/22090d695ec48bbce31d3267a7447b3f470b9188/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FTaskController.java)|78.07% **`-8.92%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/22090d695ec48bbce31d3267a7447b3f470b9188/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|76.87% **`-17.24%`**|:x:|
|[MinioService.java](https://github.com/itmo-events-app/event-app/blob/22090d695ec48bbce31d3267a7447b3f470b9188/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fminio%2FMinioService.java)|68.46% **`-19.14%`**|:x:|
|[TaskSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/22090d695ec48bbce31d3267a7447b3f470b9188/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FTaskSecurityExpression.java)|58.6% **`-16.67%`**|:green_apple:|

**Comment by IP-13** [comment]
|Overall Project|66.08% **`-4.26%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|30.75%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[TaskController.java](https://github.com/itmo-events-app/event-app/blob/c805d39168e2b706f23cd67aff71da0f0e8a89ea/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FTaskController.java)|78.07% **`-8.92%`**|:x:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/c805d39168e2b706f23cd67aff71da0f0e8a89ea/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|76.87% **`-17.24%`**|:x:|
|[MinioService.java](https://github.com/itmo-events-app/event-app/blob/c805d39168e2b706f23cd67aff71da0f0e8a89ea/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fminio%2FMinioService.java)|68.46% **`-19.14%`**|:x:|
|[TaskSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/c805d39168e2b706f23cd67aff71da0f0e8a89ea/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FTaskSecurityExpression.java)|61.56% **`-16.67%`**|:green_apple:|

**Comment by IP-13** [comment]
|Overall Project|66.73% **`-3.77%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|49.69%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EventService.java](https://github.com/itmo-events-app/event-app/blob/ad786ea6b21861d117528bc75ce89c62a2a209ac/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventService.java)|90.12%|:green_apple:|
|[MinioService.java](https://github.com/itmo-events-app/event-app/blob/ad786ea6b21861d117528bc75ce89c62a2a209ac/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fminio%2FMinioService.java)|78.44% **`-9.16%`**|:green_apple:|
|[TaskController.java](https://github.com/itmo-events-app/event-app/blob/ad786ea6b21861d117528bc75ce89c62a2a209ac/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FTaskController.java)|78.2% **`-13.15%`**|:green_apple:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/ad786ea6b21861d117528bc75ce89c62a2a209ac/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|78.06% **`-16.1%`**|:x:|
|[TaskSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/ad786ea6b21861d117528bc75ce89c62a2a209ac/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FTaskSecurityExpression.java)|58.6% **`-16.67%`**|:green_apple:|

File: src/main/java/org/itmo/eventapp/main/controller/TaskController.java [hunk 5]
```diff
@@ -147,11 +180,11 @@ public ResponseEntity<List<TaskResponse>> taskListMove(
             @RequestBody List<Integer> taskIds
     ) {
         List<Task> updTasks = taskService.moveTasks(dstEventId, taskIds);
-        return ResponseEntity.ok().body(TaskMapper.tasksToTaskResponseList(updTasks));
+        return ResponseEntity.ok().body(TaskMapper.tasksToTaskResponseList(updTasks, taskService));
     }
 
     @Operation(summary = "Копирование списка задач")
-    @PreAuthorize("@taskSecurityExpression.canCreateTask(#dstEventId)")
+    @PreAuthorize("@taskSecurityExpression.getCanCopyTasks(#dstEventId, #taskIds)")
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/controller/TaskController.java:187]: можно просто canCopyTasks назвать

File: src/main/java/org/itmo/eventapp/main/minio/MinioService.java [hunk 6]
```diff
@@ -58,29 +62,48 @@ public void delete(String bucket, String object) {
...
             Item item = iterator.next().get();
             minioClient.removeObject(RemoveObjectArgs.builder().bucket(bucket).object(item.objectName()).build());
         }
     }
 
+    @SneakyThrows
+    public List<String> getFileNamesByPrefix(String bucket, String prefix) {
+        List<String> filenames = new ArrayList<>();
+        boolean found = minioClient.bucketExists(BucketExistsArgs.builder().bucket(bucket).build());
+        if (!found) {
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/minio/MinioService.java:84]: Можно ошибку кидать наверно

File: src/main/java/org/itmo/eventapp/main/minio/MinioService.java [hunk 7]
```diff
@@ -58,29 +62,48 @@ public void delete(String bucket, String object) {
...
+            return filenames;
+        }
+        try {
+            Iterable<Result<Item>> results = minioClient.listObjects(ListObjectsArgs.builder().bucket(bucket).prefix(prefix).build());
+            for (Result<Item> result : results) {
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/minio/MinioService.java:89]: stream api

File: src/main/java/org/itmo/eventapp/main/minio/MinioService.java [hunk 8]
```diff
@@ -58,29 +62,48 @@ public void delete(String bucket, String object) {
...
+        }
+        return filenames;
+    }
+
     @SneakyThrows
     public void copyImagesWithPrefix(String sourceBucket, String destinationBucket, String sourcePrefix, String destinationPrefix) {
         try {
             if (!minioClient.bucketExists(BucketExistsArgs.builder().bucket(sourceBucket).build())) {
-                throw new MinioException("Source bucket does not exist");
+                minioClient.makeBucket(MakeBucketArgs.builder().bucket(sourceBucket).build());
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/minio/MinioService.java:102]: а тут ошибку надо, нельзя копировать оттуда, чего нет

File: src/main/java/org/itmo/eventapp/main/security/securityexpression/TaskSecurityExpression.java [hunk 9]
```diff
@@ -111,4 +112,39 @@ public boolean canGetTask(@Min(value = 1, message = "Параметр taskId н
...
+
+    }
+
+    public boolean getCanCopyTasks(int dstEventId, List<Integer> taskIds) {
+
+        int eventId = miscSecurityExpression.getParentEventOrSelfId(dstEventId);
+
+        Set<Integer> eventIds = taskService.findAllById(taskIds).stream().map(
+                task-> miscSecurityExpression.getParentEventOrSelfId(task.getEvent().getId())
+        ).collect(Collectors.toSet());
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/security/securityexpression/TaskSecurityExpression.java:138]: distinct().toList()

File: src/main/java/org/itmo/eventapp/main/service/TaskService.java [hunk 10]
```diff
@@ -124,10 +132,62 @@ public Task edit(Integer id, TaskRequest taskRequest) {
         return newTaskData;
     }
 
+    @Transactional
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/service/TaskService.java:135]: а зачем тут транзакция? Минио не бд ведь, никак не влияет

File: src/main/java/org/itmo/eventapp/main/service/TaskService.java [hunk 11]
```diff
@@ -124,10 +132,62 @@ public Task edit(Integer id, TaskRequest taskRequest) {
...
+
+        if (!Objects.isNull(files)) {
+
+            for (MultipartFile file : files) {
+
+                String modifiedFileName = task.getId().toString()
+                        + "_"
+                        + FilenameUtils.getBaseName(file.getOriginalFilename())
+                        + "_"
+                        + System.currentTimeMillis()
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/service/TaskService.java:157]: а может dateTime? А то сложно потом юзеру ориентироваться, а так хоть дата будет
LocalDateTime.now()

File: src/main/java/org/itmo/eventapp/main/service/TaskService.java [hunk 12]
```diff
@@ -209,10 +271,17 @@ public List<Task> copyTasks(Integer dstEventId, List<Integer> taskIds) {
             newTask.setStatus(status);
 
             newTasks.add(newTask);
+            prefixes.add(task.getId().toString());
         }
 
         newTasks = taskRepository.saveAll(newTasks);
 
+        for (int i = 0; i < newTasks.size(); i++) {
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/service/TaskService.java:279]: а что ты stream api так нелюбишь, удобная же штука

Comment (IP-13) [reply src/main/java/org/itmo/eventapp/main/service/TaskService.java:279]: Это так, лирика, не критично

File: src/main/java/org/itmo/eventapp/main/service/TaskService.java [hunk 13]
```diff
@@ -209,10 +271,17 @@ public List<Task> copyTasks(Integer dstEventId, List<Integer> taskIds) {
             newTask.setStatus(status);
 
             newTasks.add(newTask);
+            prefixes.add(task.getId().toString());
```
Comment (IP-13) [inline src/main/java/org/itmo/eventapp/main/service/TaskService.java:274]: а зачем тут у старой таски айди брать?
Нам нужно у новой, мы же к новой таске хотим файл привязать

Comment (IP-13) [reply src/main/java/org/itmo/eventapp/main/service/TaskService.java:274]: Это критично

**Review by IP-13 — APPROVED** [review]

**Comment by IP-13** [comment]
|Overall Project|66.7% **`-3.77%`**|:green_apple:|
|:-|:-|:-:|
|Files changed|49.69%|:x:|
<br>

|File|Coverage||
|:-|:-|:-:|
|[EventService.java](https://github.com/itmo-events-app/event-app/blob/8c20664a2fec798a53950236688748ba268103f8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FEventService.java)|90.12%|:green_apple:|
|[MinioService.java](https://github.com/itmo-events-app/event-app/blob/8c20664a2fec798a53950236688748ba268103f8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fminio%2FMinioService.java)|78.44% **`-9.16%`**|:green_apple:|
|[TaskController.java](https://github.com/itmo-events-app/event-app/blob/8c20664a2fec798a53950236688748ba268103f8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fcontroller%2FTaskController.java)|78.2% **`-13.15%`**|:green_apple:|
|[TaskService.java](https://github.com/itmo-events-app/event-app/blob/8c20664a2fec798a53950236688748ba268103f8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fservice%2FTaskService.java)|78.06% **`-16.1%`**|:x:|
|[TaskSecurityExpression.java](https://github.com/itmo-events-app/event-app/blob/8c20664a2fec798a53950236688748ba268103f8/src%2Fmain%2Fjava%2Forg%2Fitmo%2Feventapp%2Fmain%2Fsecurity%2Fsecurityexpression%2FTaskSecurityExpression.java)|58.6% **`-16.67%`**|:green_apple:|

**Pull request merged**